analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/TestAnalysisComparer.cs

```diff
@@ -0,0 +1,132 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.UnitTest.TestFramework
+{
+    public static class TestAnalysisComparer
+    {
+        public static StringBuilder Compare(Diagnostic[] diagnostics, Dictionary<string, IList<IIssueLocation>> expectedIssuesPerFile, string languageVersion)
+        {
+            var summary = new StringBuilder($"Comparing actual issues with expected:\nLanguage version: {languageVersion}.\n");
+            var actualIssues = diagnostics.ToIssueLocations();
+
+            foreach (var fileName in actualIssues.Keys.OrderBy(x => x))
+            {
+                summary.Append('\n').Append(fileName).Append('\n');
+
+                foreach (var actualIssue in actualIssues[fileName].OrderBy(x => x.LineNumber))
+                {
+                    var issue = expectedIssuesPerFile[fileName].FirstOrDefault(x => x.IsPrimary == actualIssue.IsPrimary && x.LineNumber == actualIssue.LineNumber);
```

Comment:
We'll need more selective (priority-based) filtering here and add UTs for those. There could be a single line with:
* Two issues of same rule (no problem)
* Two issues of different rule IDs
* Two issues on different locations (and it should not matter which one we mark first in UTs)*
* Two issues, one on wrong location, one on correct location
* Two issues, both on wrong location


*This should work exactly the same way
```
  Something(); // Noncompliant {{Some message}}
                        // Noncompliant {{Another mesage}}
```
and
```
  Something(); // Noncompliant {{Another mesage}}
                        // Noncompliant {{Some message}}
```

Reply:
Having something like this
```
Method(wrongLocation, alsoWrongLocation)
        ^^^^^^^^^^         ^^^^^^^^^^^^
```
should try to do it's best, somehow like this:
```
.OrderBy(x => Math.Abs(x.Start - actualIssue.Start))
.ThenBy(x => Math.Abs(x.Length - actualIssue.Length))
.FirstOrDefault()
```
to get the best possible approximate in case of issues (and will get the correct one in case of multiple correct issues)

Reply:
I try to do baby steps and not everything at once. This was not present in the previous implementation and I think we should add it in a different PR, when needed.